provisioning/device/samples/ProvisioningDeviceClientX509/readme.md

```diff
@@ -0,0 +1,53 @@
+# Provisioning Device Client Sample - TPM Attestation
+
+## Overview
+
+This is a quick tutorial with the steps to register a device in the Microsoft Azure IoT Hub Device Provisioning Service using X509 certificates locally generated.
+
+## How to run the sample
+
+Ensure that all prerequisite steps presented in [samples](..\) have been performed.
+
+The sample code is set up to use X509 certificates stored within a password-protected PKCS12 formatted file (certificate.pfx). To generate a self-signed certificate run the following command:
+
+`powershell .\GenerateTestCertificate.ps1`
+
+The script will prompt for a PFX password. The same password must be used when running the sample.
+
+In your Device Provisioning Service go to "Manage enrollments" and select "Individual Enrollments".
+Select "Add" then fill in the following:
+Mechanism: X.509
+Certificate: Select the public key 'certificate.cer' file.
+DeviceID: iothubx509device1 (must match the CN part of the certificate Subject)
+
+To run the sample, in command prompt enter:
```

Comment:
Why is the ID scope passed in as a command line parameter?

Reply:
Configuration should never be added to source control - promotes bad practices and enables a high risk scenario where PII or secrets get checked in by mistake.

Because in the last month we had 3 incidents where people submitted PRs containing secrets, I plan to change all samples in the future to take these via command-line parameters.

Reply:
What about just a placeholder in the source code and instruction to the user to fill it in with their own ID scope? This is a best practice for folks who will straight up copy+paste the sample code to bootstrap their own development efforts. I don't want prod devices taking ID scope as a startup parameter because it goes against some of the design principles of the DPS flow.

Reply:
> I don't want prod devices taking ID scope as a startup parameter because it goes against some of the design principles of the DPS flow.

What are the design principles that this is going against?

I see the IDScope similar to a SQL connection string. These should never be hard-coded within the source-code and are instead stored in a configuration file. 

How would one check-in the code to GitHub if it has an IDScope in it?
Also, why would a customer purchasing a software solution be required to have access to the source code and recompile it in order to use it on their devices?

Reply:
The DPS flow was designed such that the ID scope will never have to be changed once a device is first enrolled in DPS - the device may change owners or may change IoT hubs, but the ID scope does not change. The ID scope will be the same for all devices in a batch; in fact, the only differences between devices connecting to DPS are the registration ID and the TPM/cert information that is unique per device. The ultimate goal for devices registering with DPS is that they can use the exact same firmware image across a batch of devices and differ only by info stored in the on-board HSM.

Reply:
This is only a sample that can't be "burned-in" as firmware to a device. .Net (except NetMF which isn't supported for DPS) cannot run on bare-metal devices. The same exe/dll should be reusable on all devices without recompilation.
Even more, this is a sample which is meant to be run by everybody that wants a quick and safe way to try out our SDK. Requiring sample editing is neither given that the code is under git source control.